.github/workflows/build.yml

```diff
@@ -25,6 +25,10 @@ jobs:
       run: .\PCbuild\build.bat -e -p Win32
     - name: Display build info
       run: .\python.bat -m test.pythoninfo
+    - name: Install test dependencies
+      run: |
+        .\python.bat -m ensurepip --user
+        .\python.bat -m pip install --user -r Misc/requirements-test.txt
```

Comment:
Importing @zooba's comments from the compile-time-flag PR:

> As we discussed, I don't like relying on pip to install into the source tree before we've run the test suite.
>
> In general, system-wide dependencies are installed using the system package manager, or downloaded using specialised scripts and an existing Python installation (see PCbuild/get-externals.py). This would be a new network dependency, and I'd rather keep us to things that are checked in or directly grabbed from GitHub.

Reply:
My response:

I think it's particularly important that we test against the actual installed `tzdata`, particularly in the early days of the PR and *particularly* on Windows, because that's what a majority of Windows users will be using for time zone data.

I don't think we should check in a `tzdata` wheel to handle optional test dependencies, though.

If there's another mechanism for pulling in a wheel during the PR, I'm open to the idea, though if it takes too much time or effort I'd ask that we temporarily use PyPI and fix it after the main features are merged. I imagine any problems this would cause would be either immediate or very rare (after all, every other CI job for all my other projects and probably the majority of Python projects hits PyPI as part of the set-up, it's not a fly-by-night endpoint), so if we're already at the point where this is passing I'd rather accept it as a temporary additional dependency until after feature freeze when I'll have a bit more breathing room.

Reply:
I'd rather skip the tests that rely on an installed `tzdata` and test them on buildbots, where we can rely on manual configuration for stuff like this. We need to prioritise validation when it's _not_ present anyway, as that's the default case (and hence what the real majority of users will use).

If we can't mock up detection testing well enough without the real package, then we should probably refactor it so we can. I couldn't easily spot which tests rely on the real installation.

Reply:
Okay, found the tests, and it seems like you've got full coverage both with and without the package, as well as a warning that the ones _with_ the package may randomly break in the future.

We don't like randomly breaking tests independent from CPython changes, so I'm _much_ more comfortable not running those tests in CI.

Reply:
I don't really see what failure modes you are worried about here.

These tests are:

1. Quick to run
2. Only rely on a single first-party package (i.e. we control the endpoint — no worry about malicious changes to the package)
3. Only rely on PyPI (which is *not* a major failure point on any of the CI systems)
4. For a feature that is guaranteed in the PEP.

It's already the case that the AP runs don't use `tzdata` (mostly because I couldn't figure out how to get it to work, but I'm fine with just not fixing that), so we have coverage of the "tzdata present" and "tzdata not present" test cases.

I think it's at least a no-brainer to include them in the coverage builds, but configuring the buildbots to use them is *much* more complicated than configuring the CI to build them.

I understand the reluctance to introduce `pip` or installed modules as a *mandatory* part of the testing, because you should be able to develop offline, but CI builds are always done online.

Reply:
> Okay, found the tests, and it seems like you've got full coverage both with and without the package, as well as a warning that the ones _with_ the package may randomly break in the future.

Not sure what you mean by "full coverage both with and without the package". This section won't get hit if we don't load the data from tzdata: https://github.com/python/cpython/pull/19909/files#diff-1885af255e0d76189739e1ca80147e3dR4-R14

Assuming that the `tzdata` for the zones is the same bytes as the embedded version, the rest of the coverage paths should be the same once the data is read in, though (though that's part of what we're testing).

> We don't like randomly breaking tests independent from CPython changes, so I'm much more comfortable not running those tests in CI.

Well, in Misc/requirements.txt I've pinned the `tzdata` version, so there's no chance of them randomly breaking in this case. That warning is to say that the tests *themselves* don't pin the data, so if you can't accept a breakage due to some change in the data you need to pin your `tzdata` version.

In the backport I don't pin the version because I want early warning of breakages, but in CPython it could possibly be more disruptive.

Reply:
That link doesn't show me which section you're referring to, unfortunately.

Configuring the buildbots is much easier - you email the buildbot maintainers and ask them to install it and put it on PYTHONPATH. The only reason that isn't trivial for them is that _we don't do it_, and so you're inventing new global state for tests that we currently don't do. And once it's there, the slope becomes really slippery. (FWIW, I'd be making exactly the same argument if you'd added it as a git submodule.) I've already caught people slipping in new build dependencies in the CI scripts (well, we caught it when it broke something, I have no idea how long it had been there).

The problem is also that you're relying on pip working with a brand new build of Python. If someone breaks pip in a PR, they'll see your install fail and no tests will run, which is not going to help them fix it.

I think the approach I'd rather see is using urllib to download the package directly, extract it to a temporary location, and update sys.path as part of test setup, then clean up during teardown (unless it's already installed, of course). That keeps all the tools within the stdlib and ensures that a failure only breaks the relevant tests and not the entire run.

Reply:
> The problem is also that you're relying on pip working with a brand new build of Python. If someone breaks pip in a PR, they'll see your install fail and no tests will run, which is not going to help them fix it.

This is a valid point, but there are some pretty simple ways around it, for example, we could make it so the `pip install` step doesn't fail the build job, or fails the build job only after the tests are run, e.g.:

```bash
# Install tzdata, continue if failing
pip install -rMisc/requirements-test.txt || :
```

Or to still count "failed to install `tzdata`" as a failure, but only after the tests run:

```bash
PIP_SUCCESS=1
pip install -rMisc/requirements-test.txt || PIP_SUCCESS=0

... # Run the tests

# Set failing exit code if pip install failed.
[[ "$PIP_SUCCESS" -eq 0 ]] && exit 1
```

I personally think it's a bit overkill (you could make the same argument about "what if someone breaks `urllib`"), but at least in this specific case it doesn't need to be a blocker.

Reply:
> That link doesn't show me which section you're referring to, unfortunately.

Sorry, I don't know why it changes the URL if it's not a direct link. Anyway, I'm referring to `zoneinfo._common.load_tzdata`: https://github.com/python/cpython/blob/1ce93dacec107689567090f09cb29893888e67f6/Lib/zoneinfo/_common.py#L4-L14

Reply:
> we could make it so the `pip install` step doesn't fail the build job, or fails the build job only after the tests are run,

While it sounds like a good idea, my experience has shown that it's a great way to simply never run the tests. You need your dependencies to be deterministic, and you want them to fail if they're essential to the test suite. Silently ignoring failures here means we may as well just skip the relevant tests.

You can also easily mock out enough of a directory structure to make that test pass. Anything that breaks beyond that needs to be fixed in the `tzdata` module itself, and so the tests should be run there.

Reply:
@pganssle: I suggest you to revert all CI config changes and remove test-requirements.txt for now, merge the PR without it, so we will have more time to discuss the best option to run test_zoneinfo with tzdata installed by pip.

There are other options like running pip in test.test_zoneinfo, or in test.regrtest, or in "make test" and "make buildbottest", etc. ;-)

Reply:
> While it sounds like a good idea, my experience has shown that it's a great way to simply never run the tests. You need your dependencies to be deterministic, and you want them to fail if they're essential to the test suite. Silently ignoring failures here means we may as well just skip the relevant tests.

Well, as I mentioned, we could easily fail the test suite *afterwards*.

> You can also easily mock out enough of a directory structure to make that test pass.

This would defeat the purpose of the test, in my opinion. We're not trying to test any of `ZoneInfo`'s behavior from having loaded the data, we're trying to test that loading data from the PyPI package works.

> Anything that breaks beyond that needs to be fixed in the `tzdata` module itself, and so the tests should be run there.

Strong disagreement here. `tzdata` presents an interface and may have multiple consumers (`dateutil`, for example, will be a consumer). Generally speaking, you don't tests a package against the things that depend on it, you test the package's interface and anything that depends on it tests against the dependency.

We don't test that Python works with `numpy`, even though `numpy` is a huge and important consumer of our work. Maybe we *should* to decrease the feedback loop, but it's more important for `numpy` to tests against CPython than for CPython to test against `numpy`.




> @pganssle: I suggest you to revert all CI config changes and remove test-requirements.txt for now, merge the PR without it, so we will have more time to discuss the best option to run test_zoneinfo with tzdata installed by pip.
> 
> There are other options like running pip in test.test_zoneinfo, or in test.regrtest, or in "make test" and "make buildbottest", etc. ;-)

I guess I'm fine with this. I definitely was not suggesting that this needs to do the configuration long term — migrating into the make file and presumably whatever the Windows equivalent is would be better — but I did think the default should be "let's test the whole thing" rather than "let's not test it rather than have a harmless hack in the short term.

@zooba Can we compromise on including `tzdata` in the coverage builds? I think it adds no additional risk since coverage is already getting installed, and getting coverage feedback is important for ensuring that the tests are working as expected (the reference implementation has, for example, 100% coverage in the Python builds).